arbnode/simple_redis_lock.go

```diff
@@ -0,0 +1,204 @@
+	LockoutDuration time.Duration `koanf:"lockout-duration"`
+	RefreshDuration time.Duration `koanf:"refresh-duration"`
+	Key             string        `koanf:"key"`
+	BackgroundLock  bool          `koanf:"background-lock"`
+}
+
+func RedisLockConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.String(prefix+".redis-url", DefaultRedisLockConfig.RedisUrl, "URL of redis server for coordination")
+	f.String(prefix+".my-id", DefaultRedisLockConfig.RedisUrl, "this node's id prefix when acquiring the lock (optional)")
+	f.Duration(prefix+".lockout-duration", DefaultRedisLockConfig.LockoutDuration, "how long lock is held")
+	f.Duration(prefix+".refresh-duration", DefaultRedisLockConfig.RefreshDuration, "how long between consecutive calls to redis")
+	f.String(prefix+".key", prefix+"simple-lock", "key for lock")
+	f.Bool(prefix+".background-lock", DefaultRedisLockConfig.BackgroundLock, "should node always try grabing lock in background")
+}
+
+func NewSimpleRedisLock(config *SimpleRedisLockConfig, readyToLock func() bool) (*SimpleRedisLock, error) {
+	var client redis.UniversalClient
+	if config.RedisUrl == "" {
+		client = nil
+	} else {
+		redisOptions, err := redis.ParseURL(config.RedisUrl)
+		if err != nil {
+			return nil, err
+		}
+		client = redis.NewClient(redisOptions)
+	}
+	return &SimpleRedisLock{
+		myId:        config.MyId + "-" + strconv.FormatInt(rand.Int63(), 16), // unique even if config is not
+		client:      client,
+		config:      config,
+		readyToLock: readyToLock,
+	}, nil
+}
+
+var DefaultRedisLockConfig = SimpleRedisLockConfig{
+	RedisUrl:        "",
+	LockoutDuration: time.Minute,
+	RefreshDuration: time.Second * 10,
+	Key:             "",
+	BackgroundLock:  false,
+}
+
+func (l *SimpleRedisLock) attemptLock(ctx context.Context) (bool, error) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	if l.stopping || l.client == nil {
+		return false, nil
+	}
+	if !l.readyToLock() {
+		return false, nil
+	}
+	gotLock := false
+	timeAtStart := time.Now()
+
+	err := l.client.Watch(ctx, func(tx *redis.Tx) error {
+		current, err := tx.Get(ctx, l.config.Key).Result()
+		if errors.Is(err, redis.Nil) {
+			current = ""
+			err = nil
+		}
+		if err != nil {
+			return err
+		}
+		if current != "" && (current != l.myId) {
+			return nil
+		}
+		pipe := tx.TxPipeline()
+		pipe.Set(ctx, l.config.Key, l.myId, l.config.LockoutDuration)
+		pipe.PExpireAt(ctx, l.config.Key, timeAtStart.Add(l.config.LockoutDuration))
+		err = execTestPipe(pipe, ctx)
+		if errors.Is(err, redis.TxFailedErr) {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		gotLock = true
+		return nil
+	}, l.config.Key)
+
+	if !gotLock {
+		atomicTimeWrite(&l.lockedUntil, time.Time{})
+	}
+	if err != nil {
+		return false, err
+	}
+	if gotLock {
+		if l.config.BackgroundLock {
+			atomicTimeWrite(&l.lockedUntil, timeAtStart.Add(l.config.LockoutDuration))
+		} else {
+			atomicTimeWrite(&l.lockedUntil, timeAtStart.Add(l.config.RefreshDuration))
```

Comment:
We should have a small buffer (maybe a second?) so that we have time to refresh the lock before it expires. Right now I think it'll let it expire and then try to refresh it.

Reply:
My concern was that internally it'd think it no longer has the lock for a second, until it goes back to redis and re-acquires it. That's probably fine for the batch poster, but we should fix it since this is meant to be a generalized lock.

Reply:
Let's discuss this.
Generally - the lock expiring will always be asynchronous to the local node, so there must always be a spare time where local node doesn't know it has the lock though redis knows it does (better than the other way around :))